### PR TITLE
Handle sumo headers

### PR DESCRIPTION
**Pull request by sumo-drosiek**
Handle sumo headers

### CODE REVIEW

File: exporter/sumologicexporter/fields.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+	labels := make([]interface{}, 0)
+	matches := r.FindAllStringSubmatch(text, -1)
+
+	for _, matchset := range matches {
+		labels = append(labels, f[matchset[1]])
+	}
+	template := r.ReplaceAllString(text, "%s")
+
+	return fmt.Sprintf(template, labels...), nil
+}
```
Comment (sumo-drosiek) [inline exporter/sumologicexporter/fields.go:55]: I open for ideas how to optimise it

File: exporter/sumologicexporter/filter.go [hunk 2]
```diff
@@ -49,8 +43,8 @@ func newFilter(flds []string) (filter, error) {
 }
 
 // filterIn returns map of strings which matches at least one of the filter regexes
```
Comment (sumo-drosiek) [inline exporter/sumologicexporter/filter.go:45]:
```suggestion
// filterIn returns fields matches at least one of the filter regexes
```

File: exporter/sumologicexporter/filter.go [hunk 3]
```diff
@@ -64,8 +58,8 @@ func (f *filter) filterIn(attributes pdata.AttributeMap) map[string]string {
 }
 
 // filterOut returns map of strings which doesn't match any of the filter regexes
```
Comment (sumo-drosiek) [inline exporter/sumologicexporter/filter.go:60]:
```suggestion
// filterOut returns fields which doesn't match any of the filter regexes
```

File: exporter/sumologicexporter/source_format.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+	host     sourceFormat
+	category sourceFormat
+}
+
+type sourceFormat struct {
+	matches  [][]string
+	template string
+}
+
+func newSourceFormat(r *regexp.Regexp, text string) sourceFormat {
```
Comment (pmalek-sumo) [inline exporter/sumologicexporter/source_format.go:30]: Add comment what arguments does this `func` expect, i.e. what regex and what string and what are they used for.

File: exporter/sumologicexporter/filter.go [hunk 5]
```diff
@@ -48,9 +42,9 @@ func newFilter(flds []string) (filter, error) {
 	}, nil
 }
 
-// filterIn returns map of strings which matches at least one of the filter regexes
-func (f *filter) filterIn(attributes pdata.AttributeMap) map[string]string {
-	returnValue := make(map[string]string)
+// filterIn returns fields which matches at least one of the filter regexes
```
Comment (pmalek-sumo) [inline exporter/sumologicexporter/filter.go:45]:
```suggestion
// filterIn returns fields which match at least one of the filter regexes
```

File: exporter/sumologicexporter/filter.go [hunk 6]
```diff
@@ -63,9 +57,9 @@ func (f *filter) filterIn(attributes pdata.AttributeMap) map[string]string {
 	return returnValue
 }
 
-// filterOut returns map of strings which doesn't match any of the filter regexes
-func (f *filter) filterOut(attributes pdata.AttributeMap) map[string]string {
-	returnValue := make(map[string]string)
+// filterOut returns fields which doesn't match any of the filter regexes
```
Comment (pmalek-sumo) [inline exporter/sumologicexporter/filter.go:60]:
```suggestion
// filterOut returns fields which don't match any of the filter regexes
```

File: exporter/sumologicexporter/fields.go [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+	for k, v := range f {
+		rv = append(rv, fmt.Sprintf("%s=%s", k, v))
+	}
+	sort.Strings(rv)
+
+	return strings.Join(rv, ", ")
+}
+
+// format converts sourceFormat to the string.
+// it replaces %{key} with the appropriate value
```
Comment (pmalek-sumo) [inline exporter/sumologicexporter/fields.go:38]:
```suggestion
// format converts sourceFormat to string.
// It replaces %{key} with the appropriate value
```

What is "appropriate"? It's not obvious

File: exporter/sumologicexporter/fields.go [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+func (f fields) format(s sourceFormat) string {
+	labels := make([]interface{}, 0)
```
Comment (pmalek-sumo) [inline exporter/sumologicexporter/fields.go:40]:
```suggestion
	labels := make([]interface{}, 0, len(s.matches))
```

File: exporter/sumologicexporter/fields.go (hunk 7 above)
Comment (sumo-drosiek) [reply exporter/sumologicexporter/fields.go:38]: `corresponding`

Comment (sumo-drosiek) [inline exporter/sumologicexporter/fields.go]:
```suggestion
// it replaces %{key} with the corresponding value
```

File: exporter/sumologicexporter/fields.go [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+	for k, v := range f {
+		rv = append(rv, fmt.Sprintf("%s=%s", k, v))
+	}
+	sort.Strings(rv)
+
+	return strings.Join(rv, ", ")
+}
+
+// format converts sourceFormat to string.
+// It replaces %{key} with the appropriate value
```
Comment (sumo-drosiek) [inline exporter/sumologicexporter/fields.go:38]:
```suggestion
// It replaces %{key} with the corresponding value
```

File: exporter/sumologicexporter/source_format_test.go [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+import (
+	"regexp"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func getTestSourceFormat(t *testing.T, template string) sourceFormat {
+	r, err := regexp.Compile(`\%\{(\w+)\}`)
```
Comment (pmalek-sumo) [inline exporter/sumologicexporter/source_format_test.go:26]:
```suggestion
	r, err := regexp.Compile(sourceRegex)
```

File: exporter/sumologicexporter/source_format_test.go [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+
+func TestNewSourceFormat(t *testing.T) {
+	expected := sourceFormat{
+		matches: []string{
+			"test",
+		},
+		template: "%s/test",
+	}
+
+	r, err := regexp.Compile(`\%\{(\w+)\}`)
```
Comment (pmalek-sumo) [inline exporter/sumologicexporter/source_format_test.go:40]:
```suggestion
	r, err := regexp.Compile(sourceRegex)
```

**Review by pmalek-sumo — COMMENTED** [review]
Ok, other than those small suggestions I think this is quite good.
I still haven't figured out a way to change functions' signatures so that we don't have to pass `fields{}` whenever we want to send stuff without logs :(

File: exporter/sumologicexporter/source_format_test.go (hunk 10 above)
Comment (sumo-drosiek) [reply exporter/sumologicexporter/source_format_test.go:26]: It won't catch accidental change of the regex const

File: exporter/sumologicexporter/sender.go [hunk 12]
```diff
@@ -36,11 +36,12 @@ type appendResponse struct {
...
-	buffer []pdata.LogRecord
-	config *Config
-	client *http.Client
-	filter filter
-	ctx    context.Context
+	buffer  []pdata.LogRecord
+	config  *Config
+	client  *http.Client
+	filter  filter
+	ctx     context.Context
```
[the remark this answers is not in the document]
Comment (sumo-drosiek) [reply exporter/sumologicexporter/sender.go:43]: This is intended to not pass context through 5 functions, Sender works totally synchronized so context will be used same way as passing it by function signature

[the remark this answers is not in the document]
Comment (sumo-drosiek) [reply exporter/sumologicexporter/sender.go:43]: Will address it in another PR, probably with enabling logs

**Pull request merged**